### PR TITLE
Adds cross-repository blob mounting behavior

### PR DESCRIPTION
**Pull request by BrianBland**
Extends blob upload POST endpoint to support mount and from query
parameters as described in #634

### CODE REVIEW

File: registry/storage/blob_test.go [hunk 1]
```diff
@@ -320,6 +320,163 @@ func TestSimpleBlobRead(t *testing.T) {
...
+	}
+
+	switch err {
+	case distribution.ErrBlobUnknown:
+		break
+	default:
+		t.Errorf("Unexpected error type stat-ing deleted manifest: %#v", err)
+	}
+}
+
```
Comment (aaronlehmann) [inline registry/storage/blob_test.go]: If it's not too difficult, it would be great to have a test that ensures that an attempt to mount a source image outside the authorization scope fails.

File: registry/handlers/blobupload.go [hunk 2]
```diff
@@ -118,7 +118,44 @@ type blobUploadHandler struct {
 // StartBlobUpload begins the blob upload process and allocates a server-side
 // blob writer session.
 func (buh *blobUploadHandler) StartBlobUpload(w http.ResponseWriter, r *http.Request) {
+	fromRepo := r.FormValue("from")
+	mountDigest := r.FormValue("mount")
+
 	blobs := buh.Repository.Blobs(buh)
+
+	if mountDigest != "" && fromRepo != "" {
```
Comment (stevvooe) [inline registry/handlers/blobupload.go]: Could we put this in a separate method?

File: blobs.go [hunk 3]
```diff
@@ -146,6 +146,10 @@ type BlobIngester interface {
 
 	// Resume attempts to resume a write to a blob, identified by an id.
 	Resume(ctx context.Context, id string) (BlobWriter, error)
+
+	// Mount adds a blob to this service from another source repository,
+	// identified by a digest.
+	Mount(ctx context.Context, sourceRepo Repository, dgst digest.Digest) (Descriptor, error)
```
Comment (stevvooe) [inline blobs.go]: This can be a reference type.

File: registry/handlers/blobupload.go [hunk 4]
```diff
@@ -118,7 +118,44 @@ type blobUploadHandler struct {
...
+		blobURL, err := buh.urlBuilder.BuildBlobURL(buh.Repository.Name(), desc.Digest)
+		if err != nil {
+			buh.Errors = append(buh.Errors, errcode.ErrorCodeUnknown.WithDetail(err))
+			return
+		}
+
+		w.Header().Set("Location", blobURL)
+		w.Header().Set("Content-Length", "0")
+		w.Header().Set("Docker-Content-Digest", desc.Digest.String())
+		w.WriteHeader(http.StatusCreated)
```
Comment (stevvooe) [inline registry/handlers/blobupload.go]: This is common with `PutBlobUploadComplete`. Can we combine the two?

File: blobs.go (hunk 3 above)
Comment (BrianBland) [inline blobs.go]: Just trying to be consistent with `Put`

File: registry/handlers/blobupload.go (hunk 4 above)
Comment (BrianBland) [inline registry/handlers/blobupload.go]: Sounds like a reasonable approach

File: blobs.go (hunk 3 above)
Comment (stevvooe) [inline blobs.go]: @BrianBland Word.

@aaronlehmann Has refactoring PR that we may want to get in first. Up to you.

Comment (BrianBland) [inline blobs.go]: Doesn't matter too much to me, as the overlap will be pretty minimal

**Comment by stevvooe** [comment]
@BrianBland Have we updated the specification for these additions?

**Comment by RichardScothern** [comment]
`registry/api/v2/routes.go` and `docs/spec/api.md.tmpl` should be updated

File: registry/client/repository.go [hunk 5]
```diff
@@ -604,6 +607,25 @@ func (bs *blobs) Resume(ctx context.Context, id string) (distribution.BlobWriter
...
+	}
+
+	resp, err := bs.client.Post(u, "", nil)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+	defer resp.Body.Close()
+
+	if SuccessStatus(resp.StatusCode) {
+		// TODO(brianbland): Check for possible Accepted response (legacy create)
```
Comment (dmcgowan) [inline registry/client/repository.go]: We also need a mechanism here to create and get `httpBlobUpload`. Possibly cache it and have the next call to `Create` use it. Returning it as part of an error does not seem like an ideal solution for fallback.

File: registry/client/repository.go [hunk 6]
```diff
@@ -604,6 +607,25 @@ func (bs *blobs) Resume(ctx context.Context, id string) (distribution.BlobWriter
...
+		return bs.Stat(ctx, dgst)
+	}
+	return distribution.Descriptor{}, handleErrorResponse(resp)
```
Comment (dmcgowan) [inline registry/client/repository.go]: Would expect `distribution.ErrBlobUnknown` or similar error when the blob could not be mounted because it did not exist for the repo.

Comment (dmcgowan) [inline registry/client/repository.go]: May actually expect that the `Accepted` instead of the 4xx case.

**Comment by vbatts** [comment]
cc @jwhonce

File: docs/spec/api.md [hunk 7]
```diff
@@ -1023,7 +1051,7 @@ A list of methods and URIs are covered in the table below:
 |------|----|------|-----------|
 | GET | `/v2/` | Base | Check that the endpoint implements Docker Registry API V2. |
 | GET | `/v2/<name>/tags/list` | Tags | Fetch the tags under the repository identified by `name`. |
-| GET | `/v2/<name>/manifests/<reference>` | Manifest | Fetch the manifest identified by `name` and `reference` where `reference` can be a tag or digest. |
+| GET | `/v2/<name>/manifests/<reference>` | Manifest | Fetch the manifest identified by `name` and `reference` where `reference` can be a tag or digest. A `HEAD` request can also be issued to this endpoint to obtain resource information without receiving all data. |
```
Comment (stevvooe) [inline docs/spec/api.md]: This is redundant. That is the definition of a HEAD request. Would it be better to state somewhere at the start that all GET should support HEAD?

Comment (BrianBland) [inline docs/spec/api.md]: Hmm, this was already in the api descriptors file but had not been filled into the spec template. We can update this descriptor, but it probably belongs in a separate PR.

Comment (stevvooe) [inline docs/spec/api.md]: O, ick. Separate PR. This should be removed.

File: AUTHORS [hunk 8]
```diff
@@ -21,6 +21,7 @@ Ben Firshman <ben@firshman.co.uk>
 bin liu <liubin0329@gmail.com>
```
Comment (stevvooe) [inline AUTHORS]: Don't include AUTHORS update in your PR. We do this at release time.

File: registry/storage/blob_test.go (hunk 1 above)
Comment (stevvooe) [inline registry/storage/blob_test.go]: Can that be broken out into a specific test? It will need a mocked auth implementation to test this properly.

File: AUTHORS (hunk 8 above)
Comment (BrianBland) [inline AUTHORS]: :+1: removing this

File: registry/client/repository.go [hunk 9]
```diff
@@ -604,6 +607,25 @@ func (bs *blobs) Resume(ctx context.Context, id string) (distribution.BlobWriter
 	panic("not implemented")
 }
 
+func (bs *blobs) Mount(ctx context.Context, sourceRepo distribution.Repository, dgst digest.Digest) (distribution.Descriptor, error) {
```
Comment (stevvooe) [inline registry/client/repository.go]: @aaronlehmann Does this impose too much on the client to have a fully prepared repository when only the name is required?

Comment (aaronlehmann) [inline registry/client/repository.go]: @stevvooe: @BrianBland's engine side code implements a minimal `Repository` implementation that just returns the name, so it's not that burdensome. But it would be even simpler to just take the name, if that doesn't have any downsides.

File: registry/handlers/app.go [hunk 10]
```diff
@@ -697,6 +707,9 @@ func (app *App) authorized(w http.ResponseWriter, r *http.Request, context *Cont
 
 	if repo != "" {
 		accessRecords = appendAccessRecords(accessRecords, r.Method, repo)
+		if fromRepo := r.FormValue("from"); fromRepo != "" {
+			accessRecords = appendAccessRecords(accessRecords, "GET", fromRepo)
```
Comment (stevvooe) [inline registry/handlers/app.go]: Let's put a note here about why this is a `GET`.

File: registry/client/repository.go (hunk 9 above)
Comment (stevvooe) [inline registry/client/repository.go]: @aaronlehmann I see. I think the only reason this requires a fully formed `Repository` is on the registry side. I think we may be able to look it up inside of `linkedBlobStore`, so reducing this to a reference may be possible.

@BrianBland Thoughts?

Comment (BrianBland) [inline registry/client/repository.go]: Sure, if there's a clean way of doing the cross-repository lookup in `linkedBlobStore` I'm all for it

Comment (stevvooe) [inline registry/client/repository.go]: Should be doable. It will bypass middleware, but I think this is okay, since we really care about mount. Let me know if you need pointers.

**Comment by stevvooe** [comment]
After writing https://github.com/docker/distribution/pull/1269#discussion_r49024047, I realized it might be a good idea to have an event specific to the mount call on the target repository. It may be sufficient to use the url to detect this but we may want to consider adding an action.

Looking [here](https://github.com/docker/distribution/blob/master/notifications/event.go#L49), we just need to have an action for "mount" and a source indicating where the mount happened from. Optionally, it may be okay to leave these as "push" but then populate a "from" field on target.

File: registry/client/repository.go [hunk 11]
```diff
@@ -604,6 +621,45 @@ func (bs *blobs) Resume(ctx context.Context, id string) (distribution.BlobWriter
 	panic("not implemented")
 }
 
+func (bs *blobs) Mount(ctx context.Context, sourceRepo string, dgst digest.Digest) (distribution.Descriptor, error) {
```
Comment (aaronlehmann) [inline registry/client/repository.go]: I think `sourceRepo` should be `reference.Named` rather than a string. Perhaps the digest should also be incorporated into this argument (in that case, it would be `reference.Canonical`). Not sure if that makes sense here, since canonical references are usually used for identifying manifests, not blobs. @dmcgowan @stevvooe WDYT?

**Comment by RichardScothern** [comment]
LGTM

File: registry/client/repository.go (hunk 11 above)
Comment (dmcgowan) [inline registry/client/repository.go]: As is this use of `string` is consistent with the rest of the client package. I would agree though that it would be desirable to refactor the interfaces to use the `reference` package, I would not enforce that here.

Comment (aaronlehmann) [inline registry/client/repository.go]: Yeah, I got confused about this. I remembered refactoring everything to use the reference package in #1270, but it turns out that didn't get merged and I closed it until other PRs got merged. I'll reopen that PR and bring it up to date, but this PR can go in as-is first.

**Comment by dmcgowan** [comment]
LGTM

**Comment by aaronlehmann** [comment]
LGTM
